### PR TITLE
respect global strip config

### LINKED ISSUE #3325: Unrecognized build option errors when trying to use @io_bazel_rules_go//go/config:strip
### What version of rules_go are you using?

0.35.0

### What version of gazelle are you using?

0.26.0

### What version of Bazel are you using?

5.3.1

### Does this issue reproduce with the latest releases of all the above?

Yes

### What operating system and processor architecture are you using?

Linux x86-64

### What did you do?

```
❯ bazel build //... --@io_bazel_rules_go//go/config:strip
INFO: Invocation ID: 548294b5-5b52-465c-ad99-b6361c85c0af
INFO: Invocation ID: 548294b5-5b52-465c-ad99-b6361c85c0af
ERROR: Unrecognized option: @io_bazel_rules_go//go/config:strip=true
```

### What did you see instead?

`Unrecognized option` error. Other options appear to work that are documented

### PR DESCRIPTION
**Pull request by malt3**
**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**

The docs currently state:

> Strips symbols from compiled packages and linked binaries (using the ``-w`` flag). May also be set with the ``--strip`` command line option, which affects C/C++ targets, too.

This is not true. This option cannot be set at all (as far as I can tell). It is currently not exposed as a config flag and the global bazel option for stripping is not considered either.
This PR removes the `@io_bazel_rules_go//go/config:strip` config setting and always uses the global bazel setting `--strip` and `--compilation_mode` to deduce if stripping should be enabled.
As a little extension, I also changed the linker flags to also include `-s` if stripping is enabled.

**Which issues(s) does this PR fix?**

Fixes #3325

**Other notes for review**

This is one possible resolution. See also #3526 for an alternative solution that exposes this setting as a flag.

### CODE REVIEW

**Review by fmeum — APPROVED** [review]
This looks good to me, including the addition of the `-s` flag. 

Given the very terse descriptions of the `-s` and `-w` flags in the `go tool link` documentation, could you add a test that verifies that stack traces in panics still have all the expected information? 

Ideally that test would also check that stripping is enabled when it should be, which would probably have caught the typo below.

File: go/private/BUILD.bazel [hunk 1]
```diff
@@ -30,6 +30,21 @@ config_setting(
...
+config_setting(
+    name = "is_strip_always",
+    values = {"strip": "always"},
+    visibility = ["//:__pkg__"],
+)
+
+config_setting(
+    name = "is_strip_sometimes_fastbuild",
+    values = {
+        "strip": "always",
```
[the remark this answers is not in the document]
Comment (malt3) [reply go/private/BUILD.bazel]: Good catch

**Comment by malt3** [comment]
>  could you add a test that verifies that stack traces in panics still have all the expected information

Yes! I'll try to get to it very soon.

**Comment by malt3** [comment]
I implemented tests for ELF, MACH-O and PE. The checks for debug symbols in the resulting binaries are not very good and could certainly be improved if desired. I'm not sure where to go with this or what to test for in particular.

I also added tests to trigger a panic, recover from it and compare the stack trace against a known good stack trace.

File: tests/core/strip/strip_test.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+func Test(t *testing.T) {
+	for _, test := range []struct {
+		desc, cmd, target, stripFlag, compilationMode string
+		wantStrip                                     bool
+	}{
+		{
+			desc:      "run_auto",
+			cmd:       "run",
```
Comment (fmeum) [inline tests/core/strip/strip_test.go:169]: This seems to be the only value, could this be inlined?

**Review by fmeum — APPROVED** [review]
This is great test coverage. That will make it much easier to maintain this feature.

I see that the windows run is failing, but everything else looks good.

**Comment by linzhp** [comment]
Thanks for letting me know. We can pull this into Uber's Go repo and test it

**Comment by fmeum** [comment]
We should definitely mention this in the release notes. Since we are preserving debug information and `-c opt` builds with default settings aren't affected, I would have said that's sufficient, but would like to hear what @linzhp and @achew22 think.

**Review by fmeum — APPROVED** [review]

**Comment by linzhp** [comment]
Should we also pass `-dwarf=false` to the Go compiler when `go.mode.strip` is True?

**Comment by malt3** [comment]
I'm not sure I can make a statement on the behavior with regards to the bazel cache when setting `-dwarf=false`.
Or to rephrase:
Could this lead to missing dwarf info if I first build a stripped binary and then rebuild with debugging info? Maybe someone more experienced with the internals can comment.

This discussion has some stats on the performance impact of dwarf generation: https://github.com/golang/go/issues/38777
Not a huge difference, but not nothing either so I'd be happy to add the flag.

**Review by linzhp — APPROVED** [review]
All tests passed in Uber, after some fixes removing the usage of the old strip flag, and using transitions to dbg mode when debug symbols are needed.

**Pull request merged**